File: reference/original-java/src/hammerwatchgen/Doodad.java
package hammerwatchgen;

import java.util.ArrayList;

public class Doodad extends XMLObject {

    static ArrayList<Doodad> doodads = new ArrayList<>();
    
    public enum DoodadType {
        
        VendorMisc( "doodads/special/vendor_misc.xml", 0, 0, 0 ),
        VendorCombo( "doodads/special/vendor_combo.xml", 0, 0, 0 ),
        VendorOffense( "doodads/special/vendor_offense.xml", 0, 0, 0 ),
        VendorDefense( "doodads/special/vendor_defense.xml", 0, 0, 0 ),
        Spawn( "doodads/generic/marker_spawn.xml", 1, 1, 0 ),
        ExitMarker( "doodads/generic/marker_exit.xml", 0, 0, 0 ),
        CornerLD( "doodads/theme_%s/%s_crn_l_dn.xml", 0, 2, 2 ),
        CornerLU( "doodads/theme_%s/%s_crn_l_up.xml", 0, 1, 2 ),
        CornerRD( "doodads/theme_%s/%s_crn_r_dn.xml", 0, 2, 2 ),
        CornerRU( "doodads/theme_%s/%s_crn_r_up.xml", 0, 1, 2 ),
        ExitDn( "doodads/theme_%s/%s_exit_h_dn.xml", 0, 0, 2 ),
        ExitUp( "doodads/theme_%s/%s_exit_h_up.xml", 0, 0, 2 ),
        Horizontal( "doodads/theme_%s/%s_h_8.xml", 0, 2, 2 ),
        Vertical( "doodads/theme_%s/%s_v_8.xml", 0, 1, 2 ),
        Cover( "doodads/special/color_theme_%s_16.xml", 0.5f, 0.5f, 1 ),
        Torch( "doodads/generic/lamp_torch.xml", 0.5f, 1, 0 ),
        TorchOff( "doodads/generic/lamp_torch_off.xml", 0.5f, 1, 0 ),
        CrossWall( "doodads/theme_%s/%s_x_x.xml", 0, 1, 2 ),
        VCapDown( "doodads/theme_%s/%s_v_cap_dn.xml", 0, 2, 2 ),
        VCapUp( "doodads/theme_%s/%s_v_cap_up.xml", 0, 1, 2 ),
        HCapLeft( "doodads/theme_%s/%s_h_cap_l.xml", 0, 2, 2 ),
        HCapRight( "doodads/theme_%s/%s_h_cap_r.xml", 0, 2, 2 ),        
        TDown( "doodads/theme_%s/%s_x_t_dn.xml", 0, 2, 2 ),
        TUp( "doodads/theme_%s/%s_x_t_up.xml", 0, 1, 2 ),
        TLeft( "doodads/theme_%s/%s_x_t_l.xml", 0, 1, 2 ),
        TRight( "doodads/theme_%s/%s_x_t_r.xml", 0, 1, 2 );
        public String string;
        public float yOffset;
        public float xOffset;
        public int themeSubs;
        DoodadType( String name, float x, float y, int subs ) {
            
            this.string = name;
            xOffset = x;
            yOffset = y;
            themeSubs = subs;
        }
    }
    
    float x;
    float y;
    DoodadType type;
    int id;
    String theme;
    public static void Clear() {
        
        doodads.clear();
    }
    
    public static void Delete( Doodad d ) {
        
        doodads.remove( d );
    }
    
    public static Doodad Create( float x, float y, DoodadType type, String theme ) {
        
        Doodad d = new Doodad( x, y, type, theme );
        doodads.add( d );
        return d;
    }
    
    Doodad( float x, float y, DoodadType type, String theme ) {
        
        id = Level.idCounter;
        Level.idCounter++;
        this.x = x;
        this.y = y;
        this.type = type;
        this.theme = theme;
    }
    
    public String getXML() {
        
        // create XML structure
        XMLInt idInt = new XMLInt( "id", id );
        XMLString typeString = new XMLString( "type",getDoodadString( type, theme ) );
        XMLFloat xFloat = new XMLFloat( "x", x + type.xOffset );
        XMLFloat yFloat = new XMLFloat( "y", y + type.yOffset );
        XMLBool syncBool = new XMLBool( "need-sync", false );
        
        XMLDictionary doodadDict = new XMLDictionary( "" );
        doodadDict.addData(idInt);
        doodadDict.addData(typeString);
        doodadDict.addData(xFloat);
        doodadDict.addData(yFloat);
        doodadDict.addData(syncBool);
        
        return doodadDict.getXML();
    }
    
    public String getDoodadString( DoodadType type, String theme ) {
        
        switch( type.themeSubs ) {
            case 1:
                return String.format( type.string, theme );
                
            case 2:
                return String.format( type.string, theme, theme );
                
            default:
                return type.string;
        }
    }
    
}

File: src/generator/objects/doodad.ts
import { XMLBool, XMLDictionary, XMLFloat, XMLInt, XMLObject, XMLString } from '../xml'
import { getTheme } from '../config/themes'
import type { GenerationContext } from '../core/context'

interface DoodadTypeDef {
  /** path template; %s slots are substituted with the theme letter */
  path: string
  xOffset: number
  yOffset: number
  /** how many times the theme letter is substituted into the path */
  themeSubs: 0 | 1 | 2
}

/** Wall pieces, torches, vendors, exits… (ported from Doodad.java's DoodadType enum). */
export const DoodadType = {
  VendorMisc: { path: 'doodads/special/vendor_misc.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  VendorCombo: { path: 'doodads/special/vendor_combo.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  VendorOffense: { path: 'doodads/special/vendor_offense.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  VendorDefense: { path: 'doodads/special/vendor_defense.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  Spawn: { path: 'doodads/generic/marker_spawn.xml', xOffset: 1, yOffset: 1, themeSubs: 0 },
  ExitMarker: { path: 'doodads/generic/marker_exit.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  // The animated portal to the boss prep room, as used by the authored
  // test_boss_prep_room.xml and the shipped campaign's level_11. Ships its own
  // open/closed states (defaulting to open) and two collision posts the player
  // walks between — unlike ExitMarker above, which is a flat editor decal on
  // layer -5 and is only ever laid *under* real art. [VERIFIED] 2026-08-11
  BossPortal: { path: 'doodads/generic/exit_teleport_boss.xml', xOffset: 0, yOffset: 0, themeSubs: 0 },
  CornerLD: { path: 'doodads/theme_%s/%s_crn_l_dn.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  CornerLU: { path: 'doodads/theme_%s/%s_crn_l_up.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  CornerRD: { path: 'doodads/theme_%s/%s_crn_r_dn.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  CornerRU: { path: 'doodads/theme_%s/%s_crn_r_up.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  ExitDn: { path: 'doodads/theme_%s/%s_exit_h_dn.xml', xOffset: 0, yOffset: 0, themeSubs: 2 },
  ExitUp: { path: 'doodads/theme_%s/%s_exit_h_up.xml', xOffset: 0, yOffset: 0, themeSubs: 2 },
  Horizontal: { path: 'doodads/theme_%s/%s_h_8.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  Vertical: { path: 'doodads/theme_%s/%s_v_8.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  Cover: { path: 'doodads/special/color_theme_%s_16.xml', xOffset: 0.5, yOffset: 0.5, themeSubs: 1 },
  // A floor plate the party steps on. Purely decorative by itself — it carries
  // no trigger of its own, so the rig that uses it lays a RectangleShape over
  // the art and hangs an AreaTrigger off that. Mind the two anchors: this
  // position is where the art's top-left corner lands, while a RectangleShape's
  // is its centre, so a 1x1 box covering the plate goes half a tile further on
  // (see map/buttonSeal.ts). [EMITTED] — the path is taken from a hand-edited
  // level6 that loaded.
  TriggerButton: { path: 'doodads/special/trigger_button_floor.xml', xOffset: 0.5, yOffset: 0.5, themeSubs: 0 },
  Torch: { path: 'doodads/generic/lamp_torch.xml', xOffset: 0.5, yOffset: 1, themeSubs: 0 },
  TorchOff: { path: 'doodads/generic/lamp_torch_off.xml', xOffset: 0.5, yOffset: 1, themeSubs: 0 },
  CrossWall: { path: 'doodads/theme_%s/%s_x_x.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  VCapDown: { path: 'doodads/theme_%s/%s_v_cap_dn.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  VCapUp: { path: 'doodads/theme_%s/%s_v_cap_up.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  HCapLeft: { path: 'doodads/theme_%s/%s_h_cap_l.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  HCapRight: { path: 'doodads/theme_%s/%s_h_cap_r.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  TDown: { path: 'doodads/theme_%s/%s_x_t_dn.xml', xOffset: 0, yOffset: 2, themeSubs: 2 },
  TUp: { path: 'doodads/theme_%s/%s_x_t_up.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  TLeft: { path: 'doodads/theme_%s/%s_x_t_l.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  TRight: { path: 'doodads/theme_%s/%s_x_t_r.xml', xOffset: 0, yOffset: 1, themeSubs: 2 },
  // free-standing arena cover, not a wall-band piece — the pattern matcher
  // never places it. Default resolves correctly for the classic lettered
  // themes (doodadToken === the theme id); theme h and the bonus themes
  // override it in themes.ts because their filenames don't fit this template.
  Pillar: { path: 'doodads/theme_%s/%s_special_pillar.xml', xOffset: 0, yOffset: 0, themeSubs: 2 }
} as const satisfies Record<string, DoodadTypeDef>

export type DoodadTypeName = keyof typeof DoodadType

export function doodadPath(type: DoodadTypeName, theme: string): string {
  const themeDef = getTheme(theme)

  // a theme that does not ship this piece can point at a complete replacement
  // path, which is used verbatim — no %s substitution
  const override = themeDef?.doodadOverrides?.[type]?.path
  if (override !== undefined) return override

  const token = themeDef?.doodadToken ?? theme
  const def = DoodadType[type]
  switch (def.themeSubs) {
    case 1:
      return def.path.replace('%s', token)
    case 2:
      return def.path.replace('%s', token).replace('%s', token)
    default:
      return def.path
  }
}

/**
 * Where this piece sits relative to its tile.
 *
 * The defaults in `DoodadType` compensate for the anchor of the classic themes'
 * art (`yOffset` = the asset's `<origin>` y / 16). A theme whose art is anchored
 * differently overrides them — and must, since the offset moves the doodad's
 * collision polygon, not just its sprite.
 */
export function doodadOffset(type: DoodadTypeName, theme: string): { x: number; y: number } {
  const def = DoodadType[type]
  const override = getTheme(theme)?.doodadOverrides?.[type]
  return {
    x: override?.xOffset ?? def.xOffset,
    y: override?.yOffset ?? def.yOffset
  }
}

export class Doodad extends XMLObject {
  id: number
  /**
   * Whether this doodad must be network-synced. Stock doodads never need it
   * (`false`), but a doodad that a `DestroyObject` node can later remove —
   * the boss arena's alcove seals are the only current case — must be
   * syncable so its destruction replicates ([VERIFIED] see boss-tab.md).
   *
   * A mutable field with a default, not a constructor parameter: `create` is
   * called positionally from ~15 sites, and inserting a param there would
   * silently mis-bind one of the existing arguments at every call site. A
   * caller that needs `true` sets `seal.needSync = true` after construction.
   */
  needSync = false

  constructor(
    ctx: GenerationContext,
    public x: number,
    public y: number,
    public type: DoodadTypeName,
    public theme: string
  ) {
    super()
    this.id = ctx.idCounter++
  }

  static create(ctx: GenerationContext, x: number, y: number, type: DoodadTypeName, theme: string): Doodad {
    const d = new Doodad(ctx, x, y, type, theme)
    ctx.doodads.push(d)
    return d
  }

  getXML(): string {
    const offset = doodadOffset(this.type, this.theme)
    const dict = new XMLDictionary('')
    dict.addData(new XMLInt('id', this.id))
    dict.addData(new XMLString('type', doodadPath(this.type, this.theme)))
    dict.addData(new XMLFloat('x', this.x + offset.x))
    dict.addData(new XMLFloat('y', this.y + offset.y))
    dict.addData(new XMLBool('need-sync', this.needSync))
    return dict.getXML()
  }
}
